Sources/NIO/BaseSocket.swift

```diff
@@ -58,9 +58,16 @@ import CNIOWindows
 #endif
 
 /// A Registration on a `Selector`, which is interested in an `SelectorEventSet`.
+/// `selectableSequenceIdentifier` is used by some event notification backends (io_uring)
+/// to markup outbound events to allow for filtering of asynchronously received return values to not
+/// be delivered to a new Selectable instance that receives the same handle (fd). Ok if it wraps.
+/// Needed for i.e. testWeDoNotDeliverEventsForPreviouslyClosedChannels to succeed.
+typealias SelectableSequenceIdentifier = UInt32
```

Comment:
hmm, is there really no other way to do this? If no, can we hide that inside of the uring implementation?

Reply:
Well, it is really data tagged to the registration with the exact same lifetime - I wanted to avoid additional memory allocations which was why I extended the actual Registration. If hiding it in uring, we would need to maintain a shadow/parallel dictionary with the sequence number (basically doubling the registration overhead in allocations), so this seemed like the more elegant solution that considered. (I'm assuming the comment is for the overall tagging of registrations with a sequence identifier, not the type alias itself)

Reply:
Isn't this a common problem with uring that there may be notifications about file descriptors that have been closed and whose number then got re-used?

Reply:
Yes, that's exactly what this helps protecting against - we can then check the current sequence number for the registration and compare with the one coming back from the event - so we are 100% sure that the event was originated from that specific registration.

Reply:
"current" sequence number was poorly worded. Just "sequence number" - it is immutable for the registration during its lifetime.

Reply:
@hassila This doesn't feel right to me which is why I'm a bit reluctant to take this. What I don't understand is why we can't just discard the completion queue if there was a deregistration like we do for kqueue/epoll and you'd also do for select/poll. Given that in this version we're providing a new eventing mechanism (and not the AIO backend) I don't get why this has to change.

Reply:
I'll try to explain, please ask if unclear:

There is a fundamental issue - even if just discarding the whole inbound completion queue, you can race with the kernel and get an update due to the truly asynchronous nature of `io_uring` (there might be `SQE`:s in flight to the kernel and it might just have written a `CQE` to us, but not done the memory barrier yet for visibility). 

This works for epoll (and presumably kqueue) as we basically synchronously make the modification of the poll mask `epoll_ctl` and then throw away the rest of the events, so you have an implicit synchronisation point visavi the kernel there, its not truly async like io_uring is.

The current SwiftNIO logic expects that you can throw away all notifications _during a single event loop tick_ (and that depends on that implicit synchronisation point). That is not good enough with a truly asynchronous interface with the kernel, which is why we failed the test without using the markup. The markup guarantees that we can handle async replies over any number of event loop ticks robustly.

Another secondary related issue is the ET nature of io_uring — throwing away all pending notifications would basically force us to reissue a kernel prod for each and every event (which would at least give us updates for the new registration, even if duplicate) — which could possibly trigger the `assert` in `BaseSocketChannel` discussed elsewhere.

Reply:
Ok, thanks, this explanation makes sense.

Regarding the synchronisation I assumed we can just put an event that always fires in the submission queue and discard the completion queue until we see that pseudo event, no? That way we should be able to synchronise with the kernel if necessary. Said that it make more sense to finally fix #381 and remove the need to throw away all events...

But even that would only solve half the issue, because io_uring's poll mode is always edge triggered. But let's discuss that another time.

Reply:
@weissi not even such luck, the completions arrive out-of-order without any guarantees - see the note about the `fsync` implementation [using IOSQE_IO_DRAIN](https://lwn.net/Articles/785690/) at LWN. 

The only way to truly guarantee that we sync up would be to use IOSQE_IO_DRAIN for every deregistration, it would have truly bad performance implications, then I think we better not bother rolling this in even, I believe it would be unusable in practice.

With regard to #381, tagging things with sequence number could solve this for all backends - it could be done for kqueue using the `void      *udata` in the `kevent` and `epoll_data_t data` in `struct epoll_event`. It should be fairly easy to add support for that in both and be able to solve #381 for all backends that way I think.

Reply:
That could also be broken out into a separate PR in that case I think.

Reply:
I'm also happy to provide a PR for that on Monday if you prefer to split to work. 100% up to you.

And thank you, this is super valuable!

Reply:
Thanks @weissi , have a look at https://github.com/apple/swift-nio/pull/1799 :-)